main/lsp/LSPTypechecker.cc

```diff
@@ -390,141 +331,214 @@ LSPTypechecker::SlowPathResult LSPTypechecker::runSlowPath(LSPFileUpdates &updat
 
-                {
-                    Timer timeit(config->logger, "reIndexFromFileSystem");
-
-                    auto inputFiles = pipeline::reserveFiles(*finalGS, config->opts.inputFileNames);
-                    this->indexed.clear();
-                    this->indexed.resize(finalGS->filesUsed());
-
-                    auto asts = hashing::Hashing::indexAndComputeFileHashes(*finalGS, config->opts, *config->logger,
-                                                                            absl::Span<core::FileRef>(inputFiles),
-                                                                            workers, ownedKvstore);
-                    ENFORCE(asts.hasResult(), "LSP initialization does not support cancellation");
-                    // asts are in fref order, but we (currently) don't index and compute file hashes for payload files,
-                    // so vector index != FileRef ID. Fix that by slotting them into `indexed`.
-                    for (auto &ast : asts.result()) {
-                        int id = ast.file.id();
-                        ENFORCE_NO_TIMER(id < this->indexed.size());
-                        this->indexed[id] = std::move(ast);
+                    this->workspaceFiles = pipeline::reserveFiles(*finalGS, config->opts.inputFileNames);
+                    break;
+                }
+
+                // Reindexing on the slow path derives the list of inputs files from the file table
+                case SlowPathMode::Cancelable: {
+                    timeit.emplace(this->config->logger, "slow_path_reindex");
+
+                    ENFORCE(updates.updatedFiles.size() == updates.updatedFileIndexes.size());
+
+                    // Move the indexed trees into our local cache for open files. We will still reindex them again to
+                    // force any errors, but this is a convenient way to pre-populate our cache.
+                    if (!updates.updatedFiles.empty()) {
+                        for (auto &ast : updates.updatedFileIndexes) {
+                            if (ast.tree) {
+                                updates.updatedFinalGSFileIndexes.emplace_back(std::move(ast));
+                            }
+                        }
+
+                        updates.updatedFiles.clear();
+                        updates.updatedFileIndexes.clear();
+                    }
+
+                    this->workspaceFiles.clear();
+                    this->workspaceFiles.reserve(finalGS->filesUsed());
+
+                    // Rebuild the set of filerefs we're going to index.
+                    auto ix = -1;
+                    for (const auto &file : finalGS->getFiles()) {
+                        ++ix;
+
+                        if (file == nullptr) {
+                            continue;
+                        }
+
+                        switch (file->sourceType) {
+                            case core::File::Type::NotYetRead:
```

Comment:
ooc, is it likely that we ever get here?

Reply:
I think it shouldn't be, but given that indexing would seamlessly handle loading the new file in, it felt reasonable to include it in the work list. Do you think it's worth calling it out as a separate case with an `ENFORCE`?